language-plutus-core/docs/Builtins.md

```diff
@@ -0,0 +1,340 @@
+>
+> Make a datatype that encodes the types you care about and an interpretation for these codes as types. For example, if I care about Nat and iterated Lists thereof, I might write:
+>
+> ```idris
+> data U = NAT | LIST U
+>
+> el : U -> Type
+> el NAT = Nat
+> el (LIST t) = List (el t)
+> ```
+>
+> and then I can do generic programming (that is, matching the type structure) as follows:
+>
+> ```idris
+> toString : (t : U) -> el t -> String
+> toString NAT Z = "zero"
+> toString NAT (S k) = "s(" ++ toString NAT k ++ ")"
+> toString (LIST t) [] = "nil"
+> toString (LIST t) (x::xs) = "cons(" ++ toString t x ++ ", " ++ toString (LIST t) xs ++ ")"
+> ```
+
+Another reference is the [standard Agda tutorial](www.cse.chalmers.se/~ulfn/papers/afp08/tutorial.pdf), in particular its "3.2 Universes" section.
+
+We can apply this technique in our setting. Even though Haskell lacks dependent types, it's possible to couple a tag for a type together with that type via a GADT:
+
+```haskell
+data U a where
+    UUnit :: U ()
+    UInt  :: U Int
+```
+
+`U` is a universe consisting of `UUnit` (a tag for `()`) and `UInt` (a tag for `Int`).
+
+Another example is
+
+```haskell
+data U a where
+    UBool :: U Bool
+    UList :: !(U a) -> U [a]
+```
+
+which is a universe consisting of booleans, lists of booleans, lists of lists of booleans, etc. Here are some values of that universe / the types that they encode
+
+```
+UBool               / Bool
+UList UBool         / [Bool]
+UList (UList UBool) / [[Bool]]
+```
+
+Both `U`s are of kind `* -> *` and so if we want the AST of the language to be parameterized by a set of types, we need to parameterize it by a thing of type `* -> *`. Here's what we get (ignoring built-in functions for the moment):
+
+```haskell
+data SomeType uni = forall a. SomeType (uni a)
+
+data Type uni
+    = BuiltinType (SomeType uni)
+    | <...>
+
+data SomeValueIn uni = forall a. SomeValueIn (uni a) a
+
+data Term uni
+    = Constant (SomeValueIn uni)
+    | <...>
+```
+
+So `BuiltinType` stores just a tag like before, since the actual Haskell type the tag corresponds to is ignored (`a` is bound existentially in `SomeType` and only used in `uni a`).
+
+`Constant` now stores the tag of a type as well as a value of that that type.
+
+The original hardcoded builtins can now be recovered as
+
+```haskell
+data U a where
+    UInteger :: U Integer
+
+pattern Integer = SomeType UInteger
+pattern ConstantInteger = SomeValueIn UInteger
+
+type OriginalType = Type U
+type OriginalTerm = Term U
+```
+
+Now we can parameterize the AST by different universes of types. Of course, this requires support from all procedures (type checking, evaluation, etc) and every universe must implement a certain interface (see the [default universe](https://github.com/input-output-hk/plutus/blob/master/language-plutus-core/src/Language/PlutusCore/Universe/Default.hs) for an example).
```

Comment:
Nit: I'd probably just use relative links for the links to things in the same repo.

Reply:
I wanted to make that link reference a particular git revision, so that the document bitrots slower, but failed to do that. All other links are fine, though: they reference issues, PRs and particular revisions. Fixed.